### PR TITLE
Improve parsing

### PR DESCRIPTION
**Pull request by jensdrenhaus**

### CODE REVIEW

**Comment by jensdrenhaus** [comment]
ToDo:
* Adapt CI (see tests)
* Turn ID and Auth in custom types?
* Move some validation code into the UnmarshalJSON functions

**Comment by sylv-io** [comment]
Hi @jensdrenhaus 

The test workflow if failing because of the marshal change (undefined values are now `null` and not empty strings):
```
Error:  stboot: load opts: json: cannot unmarshal string "" into Go struct field hostCfg.host_ip of type *netlink.Addr
```

 you can fix this by adapting `.github/assets/host_configuration.json`

**Comment by jensdrenhaus** [comment]
> ToDo:
> 
> * Adapt CI (see tests)
> * Wirte Marshal / Unmarshal tests for alias types of private hostCfg and not cover everything in HostCfg test.
> * Turn ID and Auth in custom types?
> * Move some validation code into the UnmarshalJSON functions

* ~~Adapt CI (see tests)~~
* ~~Wirte Marshal / Unmarshal tests for alias types of private hostCfg~~ and not cover everything in HostCfg test.
* Turn ID and Auth in custom types?
* Move some validation code into the UnmarshalJSON functions

File: opts/opts_test.go [hunk 1]
```diff
@@ -50,21 +50,28 @@ func TestNewOpts(t *testing.T) {
...
 		got, err := NewOpts(tt.loaders...)
+		assert(t, err, tt.errType, got, tt.want)
+	}
+}
 
-		if tt.errType != nil {
-			if err == nil {
-				t.Fatal("expect an error")
-			}
+func assert(t *testing.T, gotErr, wantErrType error, got, want interface{}) {
```
Comment (MDr164) [inline opts/opts_test.go:57]: Nice, I really like this approach but I feel like this function deserves its own go file as it is used in multiple test files now.

File: opts/opts_test.go [hunk 2]
```diff
@@ -50,21 +50,28 @@ func TestNewOpts(t *testing.T) {
...
+	t.Helper()
 
-			goterr, wanterr := reflect.TypeOf(err), reflect.TypeOf(tt.errType)
-			if goterr != wanterr {
-				t.Errorf("got %+v, want %+v", goterr, wanterr)
-			}
+	if wantErrType != nil {
+		if gotErr == nil {
+			t.Fatal("expect an error")
```
Comment (MDr164) [inline opts/opts_test.go:62]: Shouldn't it rather be "expected an error"?

File: opts/testdata/unmarshal/host_timestamp_bad_type.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    "network_mode":null,
+    "host_ip":null,
+    "gateway":null,
+    "dns":null,
+    "network_interface":null,
+    "provisioning_urls":null,
+    "identity":null,
+    "authentication":null,
+    "timestamp":"1970-01-01"
+}
```
Comment (MDr164) [inline opts/testdata/unmarshal/host_timestamp_bad_type.json:11]: Missing newline

File: opts/testdata/unmarshal/host_timestamp_good.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    "network_mode":null,
+    "host_ip":null,
+    "gateway":null,
+    "dns":null,
+    "network_interface":null,
+    "provisioning_urls":null,
+    "identity":null,
+    "authentication":null,
+    "timestamp":1
+}
```
Comment (MDr164) [inline opts/testdata/unmarshal/host_timestamp_good.json:11]: Missing newline

**Review by MDr164 — CHANGES_REQUESTED** [review]

File: opts/opts_test.go (hunk 2 above)
Comment (jensdrenhaus) [reply opts/opts_test.go:62]: I say `want foo`, got bar`  so let's go with present tense here, too.

**Review by MDr164 — APPROVED** [review]
So I couldn't find any remarkable mistakes or issues but that PR is fairly large so there is obviously a chance that I missed something.

**Review by MDr164 — APPROVED** [review]

**Review by sylv-io — APPROVED** [review]
Let us fix all linting stuff on a dedicated PR.

**Pull request merged**